internal/sdk/cloudian/qos.go

```diff
@@ -0,0 +1,137 @@
+	OutboundKiBsPerMin *int64
+}
+
+// nolint: gocyclo
+func (qos *QualityOfService) unmarshalJSON(raw []byte) error {
+	var data struct {
+		QOSLimitList []struct {
+			Type  string `json:"type"`
+			Value int64  `json:"value"`
+		} `json:"qosLimitList"`
+	}
+
+	if err := json.Unmarshal(raw, &data); err != nil {
+		return err
+	}
+
+	for _, item := range data.QOSLimitList {
+		if item.Value < 0 {
+			continue
+		}
+
+		v := &item.Value
+		switch item.Type {
+		case "STORAGE_QUOTA_KBYTES_LH":
+			qos.Hard.StorageQuotaKiBs = v
+		case "STORAGE_QUOTA_KBYTES_LW":
+			qos.Warning.StorageQuotaKiBs = v
+		case "STORAGE_QUOTA_COUNT_LH":
+			qos.Hard.StorageQuotaCount = v
+		case "STORAGE_QUOTA_COUNT_LW":
+			qos.Warning.StorageQuotaCount = v
+		case "REQUEST_RATE_LH":
+			qos.Hard.RequestsPerMin = v
+		case "REQUEST_RATE_LW":
+			qos.Warning.RequestsPerMin = v
+		case "DATAKBYTES_IN_LH":
+			qos.Hard.InboundKiBsPerMin = v
+		case "DATAKBYTES_IN_LW":
+			qos.Warning.InboundKiBsPerMin = v
+		case "DATAKBYTES_OUT_LH":
+			qos.Hard.OutboundKiBsPerMin = v
+		case "DATAKBYTES_OUT_LW":
+			qos.Warning.OutboundKiBsPerMin = v
+		}
+	}
+	return nil
+}
+
+// CreateQuota sets the QoS limits for a `User`. To change QoS limits, a delete and recreate is necessary.
+func (client Client) CreateQuota(ctx context.Context, user User, qos QualityOfService) error {
+	rawParams := map[string]*int64{
+		"hlStorageQuotaKBytes": qos.Hard.StorageQuotaKiBs,
+		"wlStorageQuotaKBytes": qos.Warning.StorageQuotaKiBs,
+		"hlStorageQuotaCount":  qos.Hard.StorageQuotaCount,
+		"wlStorageQuotaCount":  qos.Warning.StorageQuotaCount,
+		"hlRequestRate":        qos.Hard.RequestsPerMin,
+		"wlRequestRate":        qos.Warning.RequestsPerMin,
+		"hlDataKBytesIn":       qos.Hard.InboundKiBsPerMin,
+		"wlDataKBytesIn":       qos.Warning.InboundKiBsPerMin,
+		"hlDataKBytesOut":      qos.Hard.OutboundKiBsPerMin,
+		"wlDataKBytesOut":      qos.Warning.OutboundKiBsPerMin,
+	}
+
+	params := make(map[string]string, len(rawParams))
+	for key, raw := range rawParams {
+		val := int64(-1)
+		if raw != nil {
+			val = *raw
+		}
+		if val < -1 {
+			val = -1
+		}
+		params[key] = strconv.FormatInt(val, 10)
+	}
+
+	resp, err := client.newRequest(ctx).
+		SetQueryParam("userId", user.UserID).
+		SetQueryParam("groupId", user.GroupID).
+		SetQueryParams(params).
+		Post("/qos/limits")
+	if err != nil {
+		return err
+	}
+
+	switch resp.StatusCode() {
+	case 200:
+		return nil
+	default:
+		return fmt.Errorf("SET quota unexpected status: %d", resp.StatusCode())
+	}
+}
+
+func (client Client) GetQuota(ctx context.Context, user User) (*QualityOfService, error) {
+	resp, err := client.newRequest(ctx).
+		SetQueryParam("userId", user.UserID).
+		SetQueryParam("groupId", user.GroupID).
+		Get("/qos/limits")
+	if err != nil {
+		return nil, err
+	}
+
+	switch resp.StatusCode() {
+	case 200:
+		qos := QualityOfService{}
+		return &qos, qos.unmarshalJSON(resp.Body())
```

Comment:
ref https://github.com/statnett/provider-cloudian/pull/124/files#r1923351297
```suggestion
		qos := &QualityOfService{}
		return qos, qos.unmarshalJSON(resp.Body())
```